### ISSUE TITLE
Regression in mypy 0.930: object.__reduce__

### ISSUE THREAD

**Issue by dvarrazzo** [python/typeshed#6661]
In mypy 0.930 the signature for object.__reduce__ has been changed to 
```python
    def __reduce__(self) -> str | Tuple[object, ...]: ...
```

This is wrong, reading [the documentation](https://docs.python.org/3/library/pickle.html#object.__reduce__). The following valid code now is reported as error:

```python
from typing import Any, Tuple, Union


class Diagnostic:
    def __reduce__(self) -> Union[str, Tuple[Any, ...]]:
        res = super().__reduce__()
        if isinstance(res, tuple) and len(res) >= 3:
            res[2]["_info"] = 42

        return res
```

**Comment by srittau** [comment]
Yeah, the return type of `__reduce__` is documented, so we should revert this change. Cc @AlexWaygood

**Comment by AlexWaygood** [comment]
Yes, my bad. Needs to be reverted, agreed.

**Comment by hauntsaninja** [comment]
@dvarrazzo this should be fixed in mypy 0.931, even if https://github.com/python/mypy/issues/11820 isn't

**Comment by dvarrazzo** [comment]
Yes, this is fixed alright :) 

- You can see a 0.931 run failing on #11820 at https://github.com/psycopg/psycopg/runs/4741795872?check_suite_focus=true
- This run with 0.930 shows both the bugs: https://github.com/psycopg/psycopg/runs/4610348714?check_suite_focus=true
